### PR TITLE
unified error handling 1/2

### PR DESCRIPTION
**Pull request by franziskuskiefer**
This is the first PR for unifying and cleaning up the error handling.
Each module now has a `errors.rs` that defines errors for this module.
Only errors from this file are allowed to be used in functions of this
module. The `errors.rs` takes care of converting other errors to errors
of the respective module.

Because this is already pretty big I stopped to split this in two.
Not all modules are converted yet.

cc: #83

### CODE REVIEW

**Comment by franziskuskiefer** [comment]
> This PR is so big! Please, split it 😊

I guess I stopped a little too late 😬

File: src/codec.rs [hunk 1]
```diff
@@ -17,24 +17,11 @@ impl From<ConfigError> for CodecError {
...
 impl From<CodecError> for ConfigError {
     // TODO: tbd in #83, also this direction shouldn't be necessary.
     fn from(e: CodecError) -> Self {
         match e {
-            CodecError::DecodingError => ConfigError::DecodingError,
-            CodecError::EncodingError => ConfigError::DecodingError,
-            CodecError::Other => ConfigError::DecodingError,
+            _ => ConfigError::InvalidConfig,
         }
     }
```
Comment (raphaelrobert) [inline src/codec.rs:26]: Clippy warning

File: src/extensions/errors.rs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+#[derive(Debug, PartialEq, Clone, Copy)]
+#[repr(u16)]
+pub enum LifetimeExtensionError {
+    /// Invalid lifetime extensions.
+    Invalid = 0,
+
+    /// Lifetime extension is expired.
+    Expired = 1,
+}
```
Comment (raphaelrobert) [inline src/extensions/errors.rs:54]: Do we care about LifetimeExtensions that are not yet valid? (with a future time window)

File: src/extensions/errors.rs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+#[derive(Debug, PartialEq, Clone, Copy)]
+#[repr(u16)]
+pub enum CapabilitiesExtensionError {
+    /// Invalid capabilities extensions.
+    Invalid = 0,
+
+    /// Capabilities extension is missing a version field.
+    EmptyVersionsField = 1,
```
Comment (raphaelrobert) [inline src/extensions/errors.rs:63]: Could there be unsupported versions?

File: src/framing/mod.rs [hunk 4]
```diff
@@ -101,7 +93,7 @@ impl MLSPlaintext {
     pub fn as_application_message(&self) -> Result<&[u8], MLSPlaintextError> {
```
Comment (raphaelrobert) [inline src/framing/mod.rs:93]: (not the correct line)

Shouldn't `verify()` return an error as well instead of a `bool`?

File: src/group/managed_group/mod.rs [hunk 5]
```diff
@@ -1,7 +1,6 @@
-use crate::ciphersuite::*;
 use crate::codec::*;
```
Comment (raphaelrobert) [inline src/group/managed_group/mod.rs:1]: Why is this removed?

File: src/key_packages/errors.rs [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+//! `KeyPackageBundle`s.
+
+use crate::{config::ConfigError, extensions::ExtensionError};
+use std::error::Error;
+
+#[derive(Debug, PartialEq, Clone, Copy)]
+#[repr(u16)]
+pub enum KeyPackageError {
+    /// An unknown extension error occurred.
+    UnknownExtensionError = 0,
```
Comment (raphaelrobert) [inline src/key_packages/errors.rs:13]: What is the purpose of this error? It doesn't seem to be used anywhere.

**Review by raphaelrobert — CHANGES_REQUESTED** [review]
Looking good in general! I commented on some specific parts.

File: src/extensions/errors.rs (hunk 2 above)
Comment (franziskuskiefer) [reply src/extensions/errors.rs:54]: We currently don't tell the caller if an extension is expired or has another invalid date. This should get improved in future. But I didn't want to change the logic here in this PR.

File: src/extensions/errors.rs (hunk 3 above)
Comment (franziskuskiefer) [reply src/extensions/errors.rs:63]: Yep, but nothing we check right now. I updated #83 to follow up on these.

File: src/framing/mod.rs (hunk 4 above)
Comment (franziskuskiefer) [reply src/framing/mod.rs:93]: It probably should, but again a logic change I wouldn't want to do in this PR (updated #83).

File: src/group/managed_group/mod.rs (hunk 5 above)
Comment (franziskuskiefer) [reply src/group/managed_group/mod.rs:1]: Because it's one of the many warnings in here ...

File: src/key_packages/errors.rs (hunk 6 above)
Comment (franziskuskiefer) [reply src/key_packages/errors.rs:13]: It is used when the `ExtensionError` is not an `InvalidExtensionType`. It should never happen, but theoretically can because the `match` in the respective from has to be exhaustive.

File: src/codec.rs (hunk 1 above)
Comment (franziskuskiefer) [reply src/codec.rs:26]: This will be fixed in the next PR when this code is actually doing something useful.

**Comment by franziskuskiefer** [comment]
@raphaelrobert I didn't change anything, only responded to your comments. Let me know if you want me to change any of them anyway.

File: src/ciphersuite/mod.rs [hunk 7]
```diff
@@ -369,17 +349,14 @@ impl AeadKey {
         msg: &[u8],
         aad: &[u8],
         nonce: &AeadNonce,
-    ) -> Result<Vec<u8>, AEADError> {
-        let (ct, tag) = match aead_encrypt(
+    ) -> Result<Vec<u8>, AeadError> {
```
Comment (kkohbrok) [inline src/ciphersuite/mod.rs:352]: Just from a naming standpoint it's not clear why an `AeadError` would not be a `CryptoError`. Maybe convert?

File: src/extensions/errors.rs [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+            Self::InvalidExtensionType => {
+                "The requested extension type is not supported by OpenMLS."
+            }
+            Self::Capabilities(e) => match e {
+                CapabilitiesExtensionError::Invalid => "Error decoding a capabilities extensions.",
+                CapabilitiesExtensionError::EmptyVersionsField => {
+                    "The versions field in the extension is empty."
+                }
+                CapabilitiesExtensionError::UnsupportedCiphersuite => {
+                    "No supported cihpersuite in the extension."
```
Comment (kkohbrok) [inline src/extensions/errors.rs:110]:
```suggestion
                    "No supported ciphersuite in the extension."
```

File: src/extensions/errors.rs [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+                }
+            },
+            Self::KeyPackageId(e) => match e {
+                KeyPackageIdError::Invalid => "Error decoding a key package id extensions.",
```
Comment (kkohbrok) [inline src/extensions/errors.rs:114]:
```suggestion
                KeyPackageIdError::Invalid => "Error decoding a key package id extension.",
```

File: src/extensions/errors.rs [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+            },
+            Self::ParentHash(e) => match e {
+                ParentHashError::Invalid => "Error decoding a parent hash extensions.",
```
Comment (kkohbrok) [inline src/extensions/errors.rs:117]:
```suggestion
                ParentHashError::Invalid => "Error decoding a parent hash extension.",
```

File: src/extensions/errors.rs [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+            },
+            Self::RatchetTree(e) => match e {
+                RatchetTreeError::Invalid => "Error decoding a ratchet tree extensions.",
```
Comment (kkohbrok) [inline src/extensions/errors.rs:120]:
```suggestion
                RatchetTreeError::Invalid => "Error decoding a ratchet tree extension.",
```

File: src/extensions/errors.rs [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+            },
+            Self::Lifetime(e) => match e {
+                LifetimeExtensionError::Invalid => "Invalid lifetime extensions.",
```
Comment (kkohbrok) [inline src/extensions/errors.rs:123]:
```suggestion
                LifetimeExtensionError::Invalid => "Invalid lifetime extension.",
```

File: src/key_packages/errors.rs [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+            Self::InvalidLifetimeExtension => {
+                "The lifetime extension of the key package is not valid."
+            }
+            Self::InvalidSignature => "The key package signature is not valid.",
+            Self::LibraryError => "An unknown OpenMLS library error occurred.",
+            Self::DuplicateExtension => "Duplicate extensions are not allowed.",
+            Self::NoCiphersuitesSupplied => {
+                "Creating a new key package requires at least one ciphersuite."
+            }
+            Self::InvalidCapabilitiesExtension => {
```
Comment (kkohbrok) [inline src/key_packages/errors.rs:87]: I know you didn't pick the name here, but now that you've put a message next to it, I notice that the name of the error is a little more general than its description. Maybe we should change the name here to something like: `CiphersuiteMismatch` or similar?

**Review by kkohbrok — CHANGES_REQUESTED** [review]

**Comment by raphaelrobert** [comment]
> @raphaelrobert I didn't change anything, only responded to your comments. Let me know if you want me to change any of them anyway.

Thanks for documenting the missing checks.

One thing you should change for sure: You removed `pub type AppMessageReceived = fn(&ManagedGroup, &[u8], &[u8]);` in the managed API, please add it back.

Otherwise it looks good.

File: src/ciphersuite/mod.rs (hunk 7 above)
Comment (franziskuskiefer) [reply src/ciphersuite/mod.rs:352]: I don't want to wrap any crypto errors in a different crypto error (what would be the point?). `CryptoError` is really only for cases where something in the crypto library goes wrong that shouldn't go wrong. Everything else are "valid" cases that can occur.

File: src/key_packages/errors.rs (hunk 13 above)
Comment (franziskuskiefer) [reply src/key_packages/errors.rs:87]: Yeah the name isn't great. I'll change it.

**Comment by franziskuskiefer** [comment]
> One thing you should change for sure: You removed `pub type AppMessageReceived = fn(&ManagedGroup, &[u8], &[u8]);` in the managed API, please add it back.

That's still there 😮 (https://github.com/openmls/openmls/blob/franziskus/error-handling/src/group/managed_group/mod.rs#L286)

**Review by kkohbrok — APPROVED** [review]
Thanks! This should make errors quite a bit more relevant in general.

**Comment by raphaelrobert** [comment]
> > One thing you should change for sure: You removed `pub type AppMessageReceived = fn(&ManagedGroup, &[u8], &[u8]);` in the managed API, please add it back.
> 
> That's still there 😮 (https://github.com/openmls/openmls/blob/franziskus/error-handling/src/group/managed_group/mod.rs#L286)

Glitch on GH. It was marked as removed, but I couldn't comment on that specific line. Anyway... looks good then.

**Review by raphaelrobert — DISMISSED** [review]

**Review by kkohbrok — APPROVED** [review]

**Review by raphaelrobert — APPROVED** [review]
